File: app/src/main/java/fr/dmconcept/bob/models/ServoConfig.java
package fr.dmconcept.bob.models;

import java.util.Arrays;

/**
 * Created by jo on 21/02/14.
 */
public class ServoConfig {

    private static final int[] PERIPHERAL_PORTS = {
         1,  2,  3,  4,  5,  6,  7,
        10, 11, 12, 13, 14,
        18, 19, 20, 21, 22, 23, 24, 25, 26
    };

    public int port;
    public int start;
    public int end;
    public int frequency;

    public ServoConfig(int port, int start, int end, int frequency) {

        // Check that the port is a peripheral port
        assert Arrays.binarySearch(PERIPHERAL_PORTS, port) > 0;
        assert start > 1000 && start < 3000;
        assert end   > 1000 && end   < 3000;
        assert start < end;
        assert frequency == 50 || frequency == 100;

        this.port       = port;
        this.start      = start;
        this.end        = end;
        this.frequency  = frequency;
    }

}

File: bob-client/src/main/scala/fr/dmconcept/bob/client/models/ServoConfig.scala
package fr.dmconcept.bob.client.models

import ServoConfig._
import fr.dmconcept.bob.client.models.json.BobJsonProtocol.servoConfigFormat
import spray.json._

object ServoConfig {

  /**
   * The IOIO output ports
   */
  final val PERIPHERAL_PORTS = Vector(1, 2, 3, 4, 5, 6, 7, 10, 11, 12, 13, 14, 18, 19, 20, 21, 22, 23, 24, 25, 26)

  /**
   * The timings in microseconds for the min and max positions
   */
  final val TIMING_RANGE = 550 to 2500

  final val FREQUENCY = 50

  def deserialize(json: JsValue): ServoConfig = json.convertTo[ServoConfig]

}

case class ServoConfig(

  // The IOIO port
  port      : Int,

  // The servo timings for the start and end position
   timings: (Int,Int)

) {

  // Check that the port is a peripheral port
  assert(PERIPHERAL_PORTS.contains(port), "Invalid IOIO port")

  // Check that the timings are in the timing limits
  assert(TIMING_RANGE.contains(timings._1), "Start timing out of range"                           )
  assert(TIMING_RANGE.contains(timings._2), "End timing out of range"                             )
  assert(timings._1 < timings._2          , "The end timing must be greater than the start timing")

}
